### PR TITLE
feat(external-api) add command for setting camera facing mode

### PR DESCRIPTION
**Pull request by horymury**
- added command for setting the camera facing mode remotely
- enhanced toggleVideo command to optionally accept the facing mode
- fix startSilent asking for microphone permissions.

### CODE REVIEW

File: conference.js [hunk 1]
```diff
@@ -2024,6 +2034,77 @@ export default {
                             eventData
                         });
                     }
+
+                    if (eventData.name === CAPTURE_SCREENSHOT_MESSAGE) {
```
Comment (saghul) [inline conference.js:2038]: Can we please add all this logic to a middleware instead? conference.js is legacy and we don't want to enmlarge any more if we can help it.

File: conference.js [hunk 2]
```diff
@@ -2024,6 +2034,77 @@ export default {
...
+                        APP.store.dispatch(captureLargeVideoScreenshot()).then(dataURL => {
+                            if (!dataURL) {
+                                room.sendEndpointMessage(sender._id, {
+                                    name: SEND_SCREENSHOT_MESSAGE,
+                                    id: undefined
```
Comment (saghul) [inline conference.js:2043]: Why send a message at all if there is no content?

File: conference.js [hunk 3]
```diff
@@ -2024,6 +2034,77 @@ export default {
...
+                            while (currentIndex < dataURL.length) {
+                                const newIndex = currentIndex + IMAGE_CHUNK_SIZE;
+
+                                splitImage.chunks.push(dataURL.slice(currentIndex, newIndex));
+                                currentIndex = newIndex;
+                            }
+
+                            const size = splitImage.chunks.length;
+
+                            for (const [ idx, chunk ] of splitImage.chunks.entries()) {
```
Comment (saghul) [inline conference.js:2064]: I think we could simplify this by chunking and sending in a single loop. Can you give it a try?

File: modules/API/constants.js [hunk 4]
```diff
@@ -22,3 +22,8 @@ export const ENDPOINT_TEXT_MESSAGE_NAME = 'endpoint-text-message';
  * but rather allowing the estimations to take place.
  */
 export const MIN_ASSUMED_BANDWIDTH_BPS = -1;
+
+/**
+ * The payload name for remotely setting the camera facing mode message.
+ */
+export const CAMERA_FACING_MODE_MESSAGE = 'camera-facing-mode-message';
```
Comment (saghul) [inline modules/API/constants.js:29]: style: alpha-sort the constant

File: react/features/base/tracks/actions.web.ts [hunk 5]
```diff
@@ -263,3 +266,38 @@ async function _toggleScreenSharing(
...
+ */
+export function setCameraFacingMode(facingMode: string | undefined) {
+    return async (dispatch: IStore['dispatch'], getState: IStore['getState']) => {
+        const state = getState();
+
+        if (!isToggleCameraEnabled(state)) {
+            return;
+        }
+
+        if (!facingMode) {
```
Comment (saghul) [inline react/features/base/tracks/actions.web.ts:284]: 👌  for keeping backwards compatibility.

File: react/features/base/tracks/actions.any.ts [hunk 6]
```diff
@@ -864,3 +864,4 @@ export function toggleCamera() {
         await APP.conference.useVideoStream(newVideoTrack);
     };
 }
+
```
Comment (saghul) [inline react/features/base/tracks/actions.any.ts:867]: Unnecessary?

**Review by saghul — COMMENTED** [review]
Left some comments, PTAL!

File: conference.js (hunk 3 above)
Comment (horymury) [reply conference.js]: done

**Review by saghul — APPROVED** [review]

File: react/features/base/tracks/middleware.web.ts [hunk 7]
```diff
@@ -121,11 +127,40 @@ MiddlewareRegistry.register(store => next => action => {
...
 });
 
+/**
+ * Registers listener for {@link JitsiConferenceEvents.ENDPOINT_MESSAGE_RECEIVED} that
+ * will perform various chat related activities.
+ *
+ * @param {IJitsiConference} conference - The conference.
+ * @returns {void}
+ */
+function _addSetCameraFacingModeListener(conference: IJitsiConference) {
```
Comment (saghul) [inline react/features/base/tracks/middleware.web.ts:146]: Can you please move this to features/conference ? Generally speaking "base" features add blocks for other features to build on top, generally without UI, and actual features use them.

**Review by saghul — CHANGES_REQUESTED** [review]

**Review by saghul — APPROVED** [review]

**Comment by horymury** [comment]
Please Jenkins test this please

**Comment by damencho** [comment]
The grid is broken, I'm working on fixing it. I will rerun all failed tests when it is healthy.

**Comment by saghul** [comment]
Looks like this is ready to go @horymury ! 🚀

**Pull request merged**